### PR TITLE
Feature/quiz improvements

### PR DESCRIPTION
**Pull request by VSchlattinger**
- practice mode
- new scoring type: "Proportional with Penalty"
- randomize order
- back-end checks for quiz validity (not completely)
- block editing after quiz has started
- quick actions in quiz-exercises list for instructor ("Start Now", "Set Visible", "Open for Practice")
- let students open quiz before start if visible
- various bugfixes and other small improvements

### CODE REVIEW

File: src/main/java/de/tum/in/www1/exerciseapp/domain/Exercise.java [hunk 1]
```diff
@@ -24,7 +24,7 @@
     discriminatorType = DiscriminatorType.STRING
 )
 @DiscriminatorValue(value = "E")
-@Cache(usage = CacheConcurrencyStrategy.NONSTRICT_READ_WRITE)
+@Cache(usage = CacheConcurrencyStrategy.READ_WRITE)
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/exerciseapp/domain/Exercise.java]: why is this change necessary?

File: src/main/java/de/tum/in/www1/exerciseapp/domain/Result.java [hunk 2]
```diff
@@ -41,6 +41,9 @@
     @Column(name = "score")
     private Long score;
 
+    @Column(name = "rated")
+    private Boolean rated;
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/exerciseapp/domain/Result.java]: this change in the data model might be problematic for other types of exercises, please add a TODO in all other places assigned to me, that we should set this value to true or false and that we also need to check for null

File: src/main/java/de/tum/in/www1/exerciseapp/domain/scoring/ScoringStrategyFactory.java [hunk 3]
```diff
@@ -18,11 +18,17 @@ public static ScoringStrategy makeScoringStrategy(Question question) {
             if (question.getScoringType() == ScoringType.ALL_OR_NOTHING) {
                 return new ScoringStrategyMultipleChoiceAllOrNothing();
             }
+            if (question.getScoringType() == ScoringType.PROPORTIONAL_WITH_PENALTY) {
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/exerciseapp/domain/scoring/ScoringStrategyFactory.java]: should be else if

File: src/main/java/de/tum/in/www1/exerciseapp/domain/scoring/ScoringStrategyFactory.java [hunk 4]
```diff
@@ -18,11 +18,17 @@ public static ScoringStrategy makeScoringStrategy(Question question) {
...
+                return new ScoringStrategyMultipleChoiceProportionalWithPenalty();
+            }
         }
         if (question instanceof DragAndDropQuestion) {
             if (question.getScoringType() == ScoringType.ALL_OR_NOTHING) {
                 return new ScoringStrategyDragAndDropAllOrNothing();
             }
+            if (question.getScoringType() == ScoringType.PROPORTIONAL_WITH_PENALTY) {
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/exerciseapp/domain/scoring/ScoringStrategyFactory.java]: should be else if

File: src/main/java/de/tum/in/www1/exerciseapp/service/StatisticService.java [hunk 5]
```diff
@@ -120,44 +127,79 @@ public void updateStatisticsAndResults(QuizExercise quizExercise){
...
         quizPointStatisticRepository.save(quizExercise.getQuizPointStatistic());
         for (Question question : quizExercise.getQuestions()) {
-            questionStatisticRepository.save(question.getQuestionStatistic());
+            if (question.getQuestionStatistic() != null) {
+                questionStatisticRepository.save(question.getQuestionStatistic());
+            }
+        }
+    }
+
+    public boolean addAndRemoveResultToStatistics(Participation participation, Result newResult, Result oldResult) {
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/exerciseapp/service/StatisticService.java]: please call this method updateStatistics and add JavaDoc. I don't really get what this method is doing.

File: src/main/java/de/tum/in/www1/exerciseapp/web/rest/QuizExerciseResource.java [hunk 6]
```diff
@@ -77,16 +86,22 @@ public QuizExerciseResource(QuizExerciseRepository quizExerciseRepository,
             return ResponseEntity.badRequest().headers(HeaderUtil.createFailureAlert(ENTITY_NAME, "idexists", "A new quizExercise cannot already have an ID")).body(null);
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/exerciseapp/web/rest/QuizExerciseResource.java]: this method has become very complex. I suggest to move parts of it into a corresponding service and to extract multiple smaller methods out of it, each with a specific single purpose. Please apply the same idea to other complex methods in Rest Resources related to quizzes and statistics

File: src/main/java/de/tum/in/www1/exerciseapp/web/rest/ResultResource.java [hunk 7]
```diff
@@ -176,9 +176,10 @@ public ResultResource(ResultRepository resultRepository, Optional<LtiService> lt
     @PreAuthorize("hasAnyRole('USER', 'TA', 'INSTRUCTOR', 'ADMIN')")
     @Timed
     public ResponseEntity<List<Result>> getResultsForParticipation(@PathVariable Long courseId,
-                                                   @PathVariable Long exerciseId,
-                                                   @PathVariable Long participationId,
-                                                   @RequestParam(defaultValue = "true") boolean showAllResults) {
+                                                                   @PathVariable Long exerciseId,
+                                                                   @PathVariable Long participationId,
+                                                                   @RequestParam(defaultValue = "true") boolean showAllResults,
+                                                                   @RequestParam(defaultValue = "false") boolean ratedOnly) {
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/exerciseapp/web/rest/ResultResource.java]: Did you add this flag for other exercises (e.g. programming) as well?

File: src/main/webapp/i18n/de/result.json [hunk 8]
```diff
@@ -20,6 +20,7 @@
             "successful": "Successful",
             "buildArtifact": "Build Artifact",
             "score": "Score",
+            "rated": "Rated",
```
Comment (krusche) [inline src/main/webapp/i18n/de/result.json]: all translations are wrong here. Please fix them.

File: src/main/java/de/tum/in/www1/exerciseapp/domain/Result.java (hunk 2 above)
Comment (VSchlattinger) [reply src/main/java/de/tum/in/www1/exerciseapp/domain/Result.java]: Actually, if we don't change anything for the other types of Exercises, they don't need this value at all and can just leave it at null. The only place where this could be problematic is in `ResultResource.getResultsForParticipation()`, but there I added the flag `ratedOnly` and set that flag when getting rated results for quiz exercises and ignore it when getting results for other types of exercises. Everywhere else, where this value is used, is already QuizExercise specific code (e.g. the statistics).

File: src/main/java/de/tum/in/www1/exerciseapp/domain/Exercise.java (hunk 1 above)
Comment (VSchlattinger) [reply src/main/java/de/tum/in/www1/exerciseapp/domain/Exercise.java]: Turns out, its not necessary, I will revert it.

File: src/main/java/de/tum/in/www1/exerciseapp/web/rest/ResultResource.java (hunk 7 above)
Comment (VSchlattinger) [reply src/main/java/de/tum/in/www1/exerciseapp/web/rest/ResultResource.java]: In the front-end code, where this Endpoint is used, the flag is set to true for QuizExercises and false for all other types of exercises.

File: src/main/java/de/tum/in/www1/exerciseapp/web/rest/QuizExerciseResource.java (hunk 6 above)
Comment (VSchlattinger) [reply src/main/java/de/tum/in/www1/exerciseapp/web/rest/QuizExerciseResource.java]: I created QuizExerciseService.java and moved some code to that.

File: src/main/java/de/tum/in/www1/exerciseapp/service/StatisticService.java (hunk 5 above)
Comment (VSchlattinger) [reply src/main/java/de/tum/in/www1/exerciseapp/service/StatisticService.java]: @MoritzIssig will do this.

File: src/main/java/de/tum/in/www1/exerciseapp/domain/Result.java (hunk 2 above)
Comment (krusche) [reply src/main/java/de/tum/in/www1/exerciseapp/domain/Result.java]: Still we have to test that...
